python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -0,0 +1,723 @@
+# Copyright (c) 2015 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import urlparse
+import hashlib
+
+import sgtk
+from sgtk.platform.qt import QtCore
+from sgtk import TankError
+
+from .background_task_manager import BackgroundTaskManager
+
+class ShotgunDataRetriever(QtCore.QObject):
+    """
+    Asyncrounous Shotgun data retriever used to execute queries and download/manage
+    thumbnails from Shotgun. Uses the BackgroundTaskManager to run tasks in background 
+    threads and emits signals when each query has either completed or failed.
+    """
+
+    @staticmethod
+    def download_thumbnail(url, bundle):
+        """
+        Convenience and compatibility method for quick and easy synchrnous thumbnail download.
+        This will retrieve a shotgun thumbnail given a url - if it already exists in the cache,
+        a path to it will be returned instantly. If not, it will be downloaded from Shotgun,
+        placed in the standard cache location on disk and its path will be returned.
+
+        This is a helper method meant to make it easy to port over synchronous legacy
+        code - for a better solution, we recommend using the thumbnail retrieval
+        that runs in a background thread.
+
+        Bcause Shotgun thumbnail urls have an expiry time, make sure to only
+        pass urls to this method that have been very recently retrieved via a Shotgun find call.
+
+        :param url: The thumbnail url string that is associated with this thumbnail. This is
+                    the field value as returned by a Shotgun query.
+        :param bundle: App, Framework or Engine object requesting the download.
+
+        :returns: A path to the thumbnail on disk.
+        """
+
+        path_to_cached_thumb = ShotgunDataRetriever._get_thumbnail_path(url, bundle)
+
+        if not os.path.exists(path_to_cached_thumb):
+
+            # create folders on disk
+            bundle.ensure_folder_exists(os.path.dirname(path_to_cached_thumb))
+
+            # download using standard core method. This will ensure that
+            # proxy and connection settings as set in the SG API are used
+            sgtk.util.download_url(bundle.shotgun, url, path_to_cached_thumb)
+
+            # modify the permissions of the file so it's writeable by others
```

Comment:
Typically the cache is either in the users home directory (so user permissions will already apply) or in a shared network location where the client can set the permissions on the containing directory if needed.  We've had this conversation a few times in the past and agreed that as permissions can be set up for the cache directories as needed (via a core hook) that this isn't a major issue...